crates/sui-core/src/authority.rs

```diff
@@ -1098,42 +1100,63 @@ impl ExecutionState for AuthorityState {
+                    // In the worst case, the synchronizer of the consensus client will catch up.
+                    self.database.persist_certificate_and_lock_shared_objects(
+                        *certificate,
+                        execution_indices,
+                    )?;
+                }
 
-        // If we already executed this transaction, return the signed effects.
-        let digest = certificate.digest();
-        if self.database.effects_exists(digest)? {
-            debug!(tx_digest =? digest, "Shared-object transaction already executed");
-            let info = self.make_transaction_info(digest).await?;
-            return Ok(bincode::serialize(&info).expect("Failed to serialize tx info"));
-        }
+                // TODO: This return time is not ideal.
+                Ok(Vec::default())
+            }
+            ConsensusTransaction::Checkpoint(fragment) => {
+                let seq = execution_indices;
+                if let Some(checkpoint) = &self.checkpoints {
+                    checkpoint
+                        .lock()
+                        .handle_internal_fragment(seq, *fragment)
+                        .map_err(|e| SuiError::from(&e.to_string()[..]))?;
 
-        // If we didn't already assigned shared-locks to this transaction, we do it now.
-        if !self.shared_locks_exist(&certificate).await? {
-            // Check the certificate. Remember that Byzantine authorities may input anything into
-            // consensus.
-            certificate.verify(&self.committee)?;
-
-            // Persist the certificate since we are about to lock one or more shared object.
-            // We thus need to make sure someone (if not the client) can continue the protocol.
-            // Also atomically lock the shared objects for this particular transaction and
-            // increment the last consensus index. Note that a single process can ever call
-            // this function and that the last consensus index is also kept in memory. It is
-            // thus ok to only persist now (despite this function may have returned earlier).
-            // In the worst case, the synchronizer of the consensus client will catch up.
-            self.database
-                .persist_certificate_and_lock_shared_objects(certificate, execution_indices)?;
-        }
+                    // NOTE: The method `handle_internal_fragment` is idempotent, so we don't need
+                    // to persist the execution indices. If the validator crashes, this transaction
+                    // may be reprocessed again and the checkpoint store will simply ignore it.
+                }
 
-        // TODO: This return time is not ideal.
-        Ok(Vec::default())
+                // TODO: This return time is not ideal. The authority submitting the checkpoint fragment
+                // is not expecting any reply.
+                Ok(Vec::default())
```

Comment:
Actually the truth is a reply would be welcome, to signal that something has been sequenced.